llvm/docs/LangRef.rst

```diff
@@ -650,48 +650,136 @@ literal types are uniqued in recent versions of LLVM.
+
+Unlike "unstable" pointers, the bit-wise representation is stable and
+``ptrtoint(x)`` always yields a deterministic value.
+This means transformation passes are still permitted to insert new ``ptrtoint``
+instructions.
+
+Non-integral pointers with external state
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A special case of non-integral pointers is ones that include external state
+(such as implicit bounds information or a type tag) with a target-defined size.
+An example of such a type is a CHERI capability, where there is an additional
+validity bit that is part of all pointer-typed registers, but is located in
+memory at an implementation-defined address separate from the pointer itself.
+Another example would be a fat-pointer scheme where pointers remain plain
+integers, but the associated bounds are stored in an out-of-band table.
+
+The following restrictions apply to IR level optimization passes:
+
+The ``inttoptr`` instruction does not recreate the external state and therefore
+it is target dependent whether it can be used to create a dereferenceable
+pointer. In general passes should assume that the result of such an inttoptr
+is not dereferenceable. For example, on CHERI targets an ``inttoptr`` will
+yield a capability the external state (the validity tag bit) set to zero,
+which will cause any dereference to trap.
+The ``ptrtotint`` instruction also only returns the "in-band" state and omit
+all external  state.
+These two properties mean that ``inttoptr(ptrtoint(x))`` cannot be folded to
+``x`` since the ``ptrtoint`` operation does not include the external state
+needed to reconstruct the original pointer and ``inttoptr`` cannot set it.
+
+When a ``store ptr addrspace(N) %p, ptr @dst`` of such a non-integral pointers
+is performed, the external metadata is also stored to the implementation-defined
+location. Similarly, a ``%val = load ptr addrspace(N), ptr @dst`` will fetch the
+external metadata and make it available for all uses of ``%val``.
+Similarly, the ``llvm.memcpy`` and ``llvm.memmove`` intrinsics also transfer the
+external state. This is essential to allow frontends to efficiently emit of
+copies of structures containing such pointers, since expanding all these copies
+as individual loads and stores would affect compilation speed and inhibit
+optimizations.
+
+Notionally, these external bits are part of the pointer, but since
+``inttoptr`` / ``ptrtoint``` only operate on the "in-band" bits of the pointer
+and the external bits are not explicitly exposed, they are not included in the
+size specified in the :ref:`datalayout string<langref_datalayout>`.
+
+When a pointer type has external state, all roundtrips via memory must
+be performed as loads and stores of the correct type since stores of other
+types may not propagate the external data.
+Therefore it is not legal to convert an existing load/store (or a ``llvm.memcpy`` /
+``llvm.memmove`` intrinsic) of pointer types with external state to a load/store
```

Comment:
`memcpy` of pointer types feels a bit weird

Is this basically saying you're now allowed to ever split a copy into smaller copies because it might contain a pointer with external state?

Reply:
... Or at least on an architecture with at least one `e` pointer (which might mean we'll want a function for `.mayHaveExternalPointers()` on `DataLayout` or something

Reply:
Got it

I'm just worried that this forbids memcpy() expansion

Like, it's weird to be in a situation where you can't unconditionally pessimize memcpy(it* dst, i8* src, usize len) to

```
for(usize i = 0; i < len; ++i) { 
  *(dst++) = *(src++);
}
```

Reply:
Yeah this is somewhat annoying. In the downstream CHERI LLVM forks, we included an attribute on memcpy/memmove that says "this copy does not contain capabilities" and then it can be expanded to integer loads/stores stores in IR/the backend.